helm-v3/005-test.md

```diff
@@ -0,0 +1,27 @@
+# Test Command
+
+## Overview
+
+`helm test` gives a chart author the ability to define actions on a helm release which validate that the chart is running as intended. This is useful for ensuring that a `helm install` or `helm upgrade` created/modified Kubernetes resources successfully and that the consequential `helm release` is functional.
+* A **test** is defined as a command and some arguments executed from a container running in a Kubernetes Pod. If the command completes inside the container with an exit code of 0, it is successful. All other exit codes indicate the test failed.
```

Comment:
I think it would be interesting to explore defining helm tests as [Job](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/) resources in addition to or instead of Pods.

Jobs have a few advantages over pods:
1. Already provide a well-documented API to define things like retries, timeouts, and backoff limits
1. Can be used to run [multiple pods with parallelism](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#parallel-jobs)
1. [Handle external issues like node failures](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#bare-pods)

For a while now in Helm 2 hooks have been able to wait for jobs to run to completion. We could do something pretty similar for `helm test`. For cleanup, we could delete `Job` resources when appropriate and defer to k8s garbage collection for `Pod` cleanup. There is also an [alpha time to live feature in 1.12](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#ttl-mechanism-for-finished-jobs).